consumer/docs/permissions.md

```diff
@@ -0,0 +1,286 @@
+---
+layout: default
+title: Working with Site Permissions
+summary: How to receive and respond to permission requests from websites.
+nav_exclude: true
+exclude: true
+tags: [GeckoView,Gecko,mozilla,android,WebView,mobile,permissions,site,media,grant,notifications,storage,persistent storage,app]
+---
+# Working with Site Permissions
+
+When a website wants to access to a users device, it will send out a permissions request. This document will explain how to use GeckoView to receive those requests, and respond to them by granting or denying those permissions.
+
+## The Permission Delegate
+
+The way an app interacts with site permissions in GeckoView is through the [`PermissionDelegate`][1]. There are three broad categories of permission that the `PermissionDelegate` handles, Android Permissions, Content Permissions and Media Permissions. All site permissions handled by GeckoView fall into one of these three categories.
+
+To get notified about permission requests, you need to implement the `PermissionDelegate` interface:
+
+```java
+private class ExamplePermissionDelegate implements GeckoSession.PermissionDelegate {
```

Comment:
I worry about this and the rest of the code going stale. We should have some kind of plan for updating docs along with API changes.

Reply:
Is there benefit to moving the documentation to mozilla-central and pushing it over to gh-pages the same way we do with the javadoc? We could then separate the code snippits with some tests so that they compile along with `./mach build` and use something like [`jekyll-include_snippet`](https://www.rubypigeon.com/posts/testing-example-code-in-your-jekyll-posts/)  to pull them into the posts. Then when you break the code snippit, the build fails.

Worth examining more closely I think.

Reply:
Yeah I really want something like that, similar to rust docs that compile with code